### ISSUE TITLE
Using project settings to adapt USWDS

### ISSUE THREAD

**Issue by thisisdano** [uswds/uswds#2408]
Goal: A version of USWDS where the user can customize USWDS defaults with new values from the system palette, where the system can output proper sizing values in `rem` based on a user-defined base font size.

- - -

One big hairy problem with integrating USWDS into an existing project is the problem of rem sizing. If an existing project has a different base font size than the USWDS default, it affects all sizing throughout the system. If we continue to use `rem` for sizing, developers should be able to define the base font size. 

They should also be able to modify the USWDS defaults to adapt to project needs. Settings including (but limited to) font choice, spacing, and colors should be customizable from the larger system palette.

- - -

Now, the existing system settings workflow is fairly simple: include variable overrides, then include uswds.

It's more complicated if we expect users to override from an existing set of variables (the system palette), since any override variables needs to be imported _before_ a user can apply them.

So the overrides are dependent on the system palette. But what is the system palette dependent on?
- System spacing values requires the value of a base font size
- System spacing requires the value of a grid base
- System typescale requires a normalization value of the selected face to compute the output size

Tweaks to the system could address some of these. We could disallow changing the grid base (I'm in favor of this, btw.) We could disallow custom re-mapping of spacing tshirt vars and require any new necessary spacing to be drawn directly from the system palette (possible!). We could drop typescale normalization (I'm not in favor this). We could use potentially use _some other technique_ to defer setting these variables for real until later in the build, or access them in a different way, i.e.
```sass
$primary-base: (red, 50, vivid); // in settings; note: list and not var
$primary-base: (blue, 50) !default; // in defaults
color: color($primary-base); // in component sass
```

Or something else — I know there are solutions out there.

That said, I had a solution that involved four customization files, kept in `src/stylesheets/project`:
`_uswds-project-init.scss`
- base font size
- project fonts
- grid base

`_uswds-project-settings`
- all other custom settings
- colors
- spacing
- typescale
- etc etc

`_uswds-project-utilities-settings`
- custom utilities output

`_uswds-project-custom`
- custom component sass or overrides

- - -

A project team would copy these files out of the project source (like node_modules) and into their project css. Similarly with the gulpfile. The gulpfile would then use sass `includePath`s to load first from the project's css, then from the uswds css if the project css didn't exist. This looks something like 
```
    .pipe(
      sass({
        outputStyle: 'expanded',
        includePaths: [
          'path/to/project/overrides',
          'src/stylesheets/project',
          'src/stylesheets',
        ],
      })
```

Where `path/to/project/overrides/` is the path to the user's custom project settings files. It's a way to conditionally use project files if they exist. Annnnnyyyyhhooooo, this worked fine on my machine but didn't work in circle. 

Is this the best solution, probably not. But it's the solution I had working. Any fix to `includePaths` would be a nice first step, before we go through and rewrite everything to do this in a better way.

**Comment by thisisdano** [comment]
So I've been experimenting with something I've been thinking of as a "deferred assignment" technique for using maps and lists as a way to set values project values without variables. It goes kinda like this:

1. All project settings overrides would happen before loading any USWDS files. To simplify, something like:
```sass
@import project-settings
@import uswds
```

2. Project settings would be set with lists:
```sass
$primary-lighter:     red, 10;
$primary-light:       red, 30, vivid;
$primary-base:        red, 50;
$primary-vivid:       red, 50, vivid;
$primary-dark:        red, 70, vivid;
$primary-darker:      false;
```

3. Internally, these values would be compiled into a new map:
```sass
$project-colors : (
  neutral: (
    lighter: system-color($neutral-lighter),
    light:   system-color($neutral-light),
    base:    system-color($neutral-base),
    dark:    system-color($neutral-dark),
    darker:  system-color($neutral-darker),
  ),
  primary: (
    lighter: system-color($primary-lighter),
    light:   system-color($primary-light),
    base:    system-color($primary-base),
    vivid:   system-color($primary-vivid),
    dark:    system-color($primary-dark),
    darker:  system-color($primary-darker),
  ),
  secondary: (
    light:   system-color($secondary-light),
    base:    system-color($secondary-base),
    vivid:   system-color($secondary-vivid),
    dark:    system-color($secondary-dark),
  ),
  accent-warm: (
    light:   system-color($accent-warm-lighter),
    base:    system-color($accent-warm-base),
  ),
  accent-cool: (
    light:   system-color($accent-cool-lighter),
    base:    system-color($accent-cool-base),
  ),
  inverse: (
    base:   system-color($inverse-base),
    dark:   system-color($inverse-dark),
  )
);
```

4. These values use the `system-color()` function to pull values from a `$system-colors` map:
```sass
@function system-color($props...){
  @if length($props) == 0 {
    @return false;
  } @elseif type-of(nth($props,1)) == 'list' {
    $props: nth($props,1);
    @if length($props) == 2 {
      @return map-deep-get(
        $system-colors,
        nth($props, 1),
        nth($props, 2)
      );
    } @elseif length($props) == 3 {
      @return map-deep-get(
        $system-colors,
        nth($props, 1),
        nth($props, 3),
        nth($props, 2)
      );
    }
  } @elseif nth($props, 1) == false {
    @return false;
  } @elseif length($props) == 2 {
    @return map-deep-get(
      $system-colors,
      nth($props, 1),
      nth($props, 2)
    );
  } @else {
    @return map-deep-get(
      $system-colors,
      nth($props, 1),
      nth($props, 3),
      nth($props, 2)
    );
  }
}
```

```sass
$system-colors: (
  red: (
    5 :          #fcf5f5,
    10:          #f7ddda,
    20:          #f7b7ad,
    30:          #f28f88,
    40:          #ed6b61,
    50:          #d83731,
    60:          #ab3a3a,
    70:          #6e312f,
    80:          #3b2523,
    90:          #211b1b,
    vivid: (
      5:          false,
      10:         #fccec7,
      20:         #fda99c,
      30:         #ff806c,
      40:         #ff5c48,
      50:         #e92106,
      60:         #c31f0a,
      70:         #9c1503,
      80:         #631212,
      90:         false,
    ),
  ),
...
```

5. This `$project-colors` and `$project-colors` maps would be collected into a single map...
```sass
$all-colors: map-collect(
  $system-colors,
  $project-colors
);
```

6. ...That developers would access with a `color()` function in their component sass:
```sass
background-color: color(primary, dark);
```

Et voilà?

And I think we could use a similar technique for any theme settings value — which could obviate the need to have a more complicated import pipeline.

**Comment by thisisdano** [comment]
This technique results in a system outcome influenced by the Saleforce token/function technique: https://medium.com/@sitaggart/how-we-use-sass-maps-for-design-tokens-and-developer-happiness-8bae50393114

**Comment by maya** [comment]
For step #2, it seems like it's repeating `red` a bunch of times. How about using `$color-primary` instead where it's set to `red`?

**Comment by thisisdano** [comment]
Like this kind of thing as a default/example?
```sass
$color-primary:       red;
$primary-lighter:     $color-primary, 10;
$primary-light:       $color-primary, 30, vivid;
$primary-base:        $color-primary, 50;
$primary-vivid:       $color-primary, 50, vivid;
$primary-dark:        $color-primary, 70, vivid;
$primary-darker:      false;
```

**Comment by maya** [comment]
yes, that's similar how I set it up on https://codepen.io/mbenari/pen/zWBjPr

**Comment by maya** [comment]
Closed via #2393.
